File: libs/power-log-parser/src/lib/parsers/secret-destroyed-parser.ts
import { CardType, GameTag, Zone } from '@firestone-hs/reference-data';
import { ParserGameTag } from '../enums';
import { ActionParser } from '../action-parser';
import { GameEventProvider, GameEventHelper } from '../game-event';
import { ShowEntity } from '../models/action';
import { Node, TagChange } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class SecretDestroyedParser implements ActionParser {
	readonly ParserName = 'SecretDestroyedParser';

	private GameState: GameState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.PowerTaskList &&
			node.Type === TagChange &&
			(node.Object as TagChange).Name === (GameTag.ZONE as number) &&
			((node.Object as TagChange).Value === (Zone.GRAVEYARD as number) ||
				(node.Object as TagChange).Value === (Zone.REMOVEDFROMGAME as number) ||
				(node.Object as TagChange).Value === (Zone.SETASIDE as number)) &&
			this.GameState.CurrentEntities.get((node.Object as TagChange).Entity)?.GetTag(GameTag.ZONE) ===
				(Zone.SECRET as number)
		);
	}

	AppliesOnCloseNode(node: Node, stateType: StateType): boolean {
		const appliesToShowEntity =
			node.Type === ShowEntity &&
			((node.Object as ShowEntity).GetTag(GameTag.ZONE) === (Zone.GRAVEYARD as number) ||
				(node.Object as ShowEntity).GetTag(GameTag.ZONE) === (Zone.REMOVEDFROMGAME as number)) &&
			this.GameState.CurrentEntities.has((node.Object as ShowEntity).Entity) &&
			this.GameState.CurrentEntities.get((node.Object as ShowEntity).Entity)!.GetTag(GameTag.ZONE) ===
				(Zone.SECRET as number);
		return stateType === StateType.PowerTaskList && appliesToShowEntity;
	}

	CreateGameEventProviderFromNew(node: Node): GameEventProvider[] | null {
		const tagChange = node.Object as TagChange;
		const entity = this.GameState.CurrentEntities.get(tagChange.Entity)!;
		if (entity.GetTag(ParserGameTag.SECRET_HAS_TRIGGERED) === 1) {
			return null;
		}
		const cardId = entity.CardId;
		const controllerId = entity.GetEffectiveController();
		if (tagChange.Value === (Zone.SETASIDE as number) && entity.GetZone() !== (Zone.SECRET as number)) {
			return null;
		}

		const eventName = entity.GetTag(GameTag.SECRET) === 1 ? 'SECRET_DESTROYED' : 'QUEST_DESTROYED';
		return [
			GameEventProvider.Create(
				tagChange.TimeStamp,
				eventName,
				GameEventHelper.CreateProvider(eventName, cardId, controllerId, entity.Id, this.StateFacade),
				true,
				node,
			),
		];
	}

	CreateGameEventProviderFromClose(node: Node): GameEventProvider[] | null {
		const showEntity = node.Object as ShowEntity;
		const entity = this.GameState.CurrentEntities.get(showEntity.Entity);
		if (entity == null || entity.GetTag(ParserGameTag.SECRET_HAS_TRIGGERED) === 1) {
			return null;
		}

		const cardId = showEntity.CardId;
		const controllerId = showEntity.GetTag(GameTag.CONTROLLER);
		const playerClass = showEntity.GetPlayerClass();
		const eventName = showEntity.GetTag(GameTag.SECRET) === 1 ? 'SECRET_DESTROYED' : 'QUEST_DESTROYED';
		return [
			GameEventProvider.Create(
				showEntity.TimeStamp,
				eventName,
				GameEventHelper.CreateProvider(eventName, cardId, controllerId, showEntity.Entity, this.StateFacade, {
					PlayerClass: playerClass,
				}),
				true,
				node,
			),
		];
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/SecretDestroyedParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;
using System.Linq;
using System.Diagnostics;

namespace HearthstoneReplays.Events.Parsers
{
    public class SecretDestroyedParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public SecretDestroyedParser(ParserState ParserState, StateFacade facade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = facade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return stateType == StateType.PowerTaskList
                && node.Type == typeof(TagChange)
                && (node.Object as TagChange).Name == (int)GameTag.ZONE
                && ((node.Object as TagChange).Value == (int)Zone.GRAVEYARD
                    // For BG quests
                    || (node.Object as TagChange).Value == (int)Zone.REMOVEDFROMGAME
                    // For the Priest quest that splits in two
                    || (node.Object as TagChange).Value == (int)Zone.SETASIDE)
                && GameState.CurrentEntities.GetValueOrDefault((node.Object as TagChange).Entity)?.GetTag(GameTag.ZONE) == (int)Zone.SECRET;
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            // When the active player destroys a secret, it is fully revealed
            var appliesToShowEntity = node.Type == typeof(ShowEntity)
                && ((node.Object as ShowEntity).GetTag(GameTag.ZONE) == (int)Zone.GRAVEYARD
                    // BG quests
                    || (node.Object as ShowEntity).GetTag(GameTag.ZONE) == (int)Zone.REMOVEDFROMGAME)
                && GameState.CurrentEntities.ContainsKey((node.Object as ShowEntity).Entity)
                && GameState.CurrentEntities[(node.Object as ShowEntity).Entity].GetTag(GameTag.ZONE) == (int)Zone.SECRET;
            return stateType == StateType.PowerTaskList
                && appliesToShowEntity;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            var tagChange = node.Object as TagChange;
            var entity = GameState.CurrentEntities[tagChange.Entity];
            if (entity.GetTag(GameTag.SECRET_HAS_TRIGGERED) == 1)
            {
                return null;
            }
            var cardId = entity.CardId;
            var controllerId = entity.GetEffectiveController();
            if (tagChange.Value == (int)Zone.SETASIDE && entity.GetZone() != (int)Zone.SECRET)
            {
                // This is to handle the Priest quest that splits in two
                return null;
            }

            //var gameState = GameEvent.BuildGameState(ParserState, StateFacade, GameState, tagChange, null);
            var eventName = entity.GetTag(GameTag.SECRET) == 1
                ? "SECRET_DESTROYED"
                : "QUEST_DESTROYED";
            return new List<GameEventProvider> { GameEventProvider.Create(
                tagChange.TimeStamp,
                eventName,
                GameEvent.CreateProvider(
                    eventName,
                    cardId,
                    controllerId,
                    entity.Id,
                    StateFacade),
                true,
                node) };
        }

        // Typically the case when the opponent plays a quest or a secret
        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            var showEntity = node.Object as ShowEntity;
            var entity = GameState.CurrentEntities.GetValueOrDefault(showEntity.Entity);
            if (entity == null || entity.GetTag(GameTag.SECRET_HAS_TRIGGERED) == 1)
            {
                return null;
            }

            var cardId = showEntity.CardId;
            var controllerId = showEntity.GetTag(GameTag.CONTROLLER);
            //var gameState = GameEvent.BuildGameState(ParserState, StateFacade, GameState, null, showEntity);
            var playerClass = showEntity.GetPlayerClass();
            var eventName = showEntity.GetTag(GameTag.SECRET) == 1
                ? "SECRET_DESTROYED"
                : "QUEST_DESTROYED";
            return new List<GameEventProvider> { GameEventProvider.Create(
                showEntity.TimeStamp,
                eventName,
                GameEvent.CreateProvider(
                    eventName,
                    cardId,
                    controllerId,
                    showEntity.Entity,
                    StateFacade,
                    //gameState,
                    new {
                        PlayerClass = playerClass,
                    }),
                true,
                node) };
        }
    }
}
